pydantic_ai_slim/pydantic_ai/prompt_config.py

```diff
@@ -0,0 +1,414 @@
+
+    def apply_template_message_history(self, _messages: list[ModelMessage], ctx: RunContext[Any]) -> list[ModelMessage]:
+        return [
+            replace(
+                message,
+                parts=[self.apply_template(part, ctx) for part in message.parts],
+            )
+            if isinstance(message, ModelRequest)
+            else message
+            for message in _messages
+        ]
+
+    def _apply_retry_template(
+        self,
+        message_part: RetryPromptPart,
+        ctx: RunContext[Any],
+    ) -> RetryPromptPart:
+        """Render the full retry response based on content type.
+
+        Selects the appropriate templates and applies them in a single pass,
+        pre-rendering everything so model_response() can just return the result.
+        """
+        content = message_part.content
+
+        if isinstance(content, str):
+            if message_part.tool_name is None:
+                # String without tool context (e.g., output validator raising ModelRetry)
+                description_template = self.description_template or default_validation_feedback
+                description = description_template(content)
+                retry_template = self.no_tool_retry_prompt or DEFAULT_MODEL_RETRY
+            else:
+                # String from a tool - use content directly
+                description = content
+                retry_template = self.tool_retry_prompt or DEFAULT_MODEL_RETRY
+        else:
+            # List of ErrorDetails (validation errors)
+            description_template = self.description_template or default_validation_error
+            description = description_template(content)
+            retry_template = self.validation_errors_retry_prompt or DEFAULT_MODEL_RETRY
+
+        # Resolve callable if needed
+        if callable(retry_template):
+            retry_template = retry_template(message_part, ctx)
+
+        return replace(message_part, retry_message=f'{description}\n\n{retry_template}')
+
+    def _apply_tool_template(
+        self,
+        message_part: ToolReturnPart,
+        ctx: RunContext[Any],
+        template: str | Callable[[ToolReturnPart, RunContext[Any]], str],
+    ) -> ToolReturnPart:
+        content = template(message_part, ctx) if callable(template) else template
+        return replace(message_part, content=content)
+
+
+@dataclass
+class ToolConfig:
+    """Configuration for customizing tool descriptions and argument descriptions at runtime.
+
+    This allows you to override tool metadata without modifying the original tool definitions.
+
+    This is applied via [`PromptConfig`][pydantic_ai.PromptConfig] and can be used to rename tools,
+    override descriptions, enable strict schema behavior for supported models, and override argument
+    descriptions.
+    """
+
+    name: str | None = None
+    tool_description: str | None = None
+    strict: bool | None = None
+    tool_args_descriptions: dict[str, str] | None = None
```

Comment:
Trying to mirror `ToolDefinition.parameters_json_schema`, let's call this `parameters_descriptions`